pkg/validate/openshift_cluster.go

```diff
@@ -0,0 +1,216 @@
+		vlabs.Upgrading:
+	default:
+		errs = append(errs, fmt.Errorf("invalid properties.provisioningState %q", p.ProvisioningState))
+	}
+
+	switch p.OpenShiftVersion {
+	case "v3.10":
+	default:
+		errs = append(errs, fmt.Errorf("invalid properties.openShiftVersion %q", p.OpenShiftVersion))
+	}
+
+	if p.PublicHostname != "" && !isValidHostname(p.PublicHostname) {
+		errs = append(errs, fmt.Errorf("invalid properties.publicHostname %q", p.PublicHostname))
+	}
+
+	if p.FQDN != "" && !isValidHostname(p.FQDN) {
+		errs = append(errs, fmt.Errorf("invalid properties.fqdn %q", p.FQDN))
+	}
+
+	errs = append(errs, validateRouterProfiles(p.RouterProfiles)...)
+
+	errs = append(errs, validateAgentPoolProfiles(p.AgentPoolProfiles)...)
+
+	errs = append(errs, validateServicePrincipalProfile(&p.ServicePrincipalProfile)...)
+
+	return
+}
+
+func validateRouterProfiles(rps []vlabs.RouterProfile) (errs []error) {
+	rpmap := map[string]vlabs.RouterProfile{}
+
+	for _, rp := range rps {
+		if _, found := rpmap[rp.Name]; found {
+			errs = append(errs, fmt.Errorf("duplicate properties.routerProfiles %q", rp.Name))
+		}
+		rpmap[rp.Name] = rp
+
+		errs = append(errs, validateRouterProfile(rp)...)
+	}
+
+	for name := range validRouterProfileNames {
+		if _, found := rpmap[name]; !found {
+			errs = append(errs, fmt.Errorf("invalid properties.routerProfiles[%q]", name))
+		}
+	}
+
+	for name := range rpmap {
+		if _, found := validRouterProfileNames[name]; !found {
+			errs = append(errs, fmt.Errorf("invalid properties.routerProfiles[%q]", name))
+		}
+	}
+
+	return
+}
+
+func validateRouterProfile(rp vlabs.RouterProfile) (errs []error) {
+	if rp.Name == "" {
+		errs = append(errs, fmt.Errorf("invalid properties.routerProfiles[%q].name %q", rp.Name, rp.Name))
+	}
+
+	if rp.PublicSubdomain != "" && !isValidHostname(rp.PublicSubdomain) {
+		errs = append(errs, fmt.Errorf("invalid properties.routerProfiles[%q].publicSubdomain %q", rp.Name, rp.PublicSubdomain))
+	}
+
+	if rp.FQDN != "" && !isValidHostname(rp.FQDN) {
+		errs = append(errs, fmt.Errorf("invalid properties.routerProfiles[%q].fqdn %q", rp.Name, rp.FQDN))
+	}
+
+	return
+}
+
+func validateAgentPoolProfiles(apps []vlabs.AgentPoolProfile) (errs []error) {
+	appmap := map[string]vlabs.AgentPoolProfile{}
+
+	for i, app := range apps {
+		if _, found := appmap[app.Name]; found {
+			errs = append(errs, fmt.Errorf("duplicate properties.agentPoolProfiles %q", app.Name))
+		}
+		appmap[app.Name] = app
+
+		if i > 0 && app.VnetSubnetID != apps[i-1].VnetSubnetID {
+			errs = append(errs, fmt.Errorf("duplicate properties.agentPoolProfiles.vnetSubnetID %q", app.VnetSubnetID))
+		}
+
+		errs = append(errs, validateAgentPoolProfile(&app)...)
+	}
+
+	for name := range validAgentPoolProfileNames {
+		if _, found := appmap[name]; !found {
+			errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q]", name))
+		}
+	}
+
+	for name := range appmap {
+		if _, found := validAgentPoolProfileNames[name]; !found {
```

Comment:
this check could be done in the first loop so that only valid app names are in the appmap (premature optimization ftw :smile:)

Reply:
thanks ;)